Use with Typescript (#200)
vercel/swr-site: pages/docs/typescript.en-US.mdx

# TypeScript

SWR is friendly for apps which are written in TypeScript. It's type safe out of the box.

### Basic Usage

By default, SWR will also infer the type of `key` argument in `fetcher`. So you can use the preferred types of `key` and `fetcher`.

#### Basic hook

```typescript

> // `key` is inferred to be `string`
> useSWR('/api/user', key => {})
> // `key` will be inferred as { a: string; b: { c: string; d: number } }
> useSWR({ a: '1', b: { c: '3', d: 2 } }, key => {})

Reviewer: ```suggestion
// `key` is inferred to be `string`
useSWR('/api/user', key => {})
useSWR(() => '/api/user', key => {})
// `key` will be inferred as { a: string; b: { c: string; d: number } }
useSWR({ a: '1', b: { c: '3', d: 2 } }, key => {})
useSWR(() => { a: '1', b: { c: '3', d: 2 } }, key => {})
// `arg0` will be inferred as string.  `arg1` will be inferred as number
useSWR(['user', 8], (arg0, arg1) => {})
useSWR(() => ['user', 8], (arg0, arg1) => {})
```

```

> You can explicitly the types for key and fetcher arguments.

Reviewer: ```suggestion
You can explicitly specify the types for key and fetcher arguments
```

```typescript
import useSWR, { Key, Fetcher } from 'swr'

const key: Key = 'string'
const fetcher: Fetcher<string, string> = (k) => k + ':value'
useSWR(key, fetcher)
```

#### Inifinite

It's same for `swr/inifite`, you can either rely on the automatically inferring or explicitly sepicify the types.

```typescript
import { SWRInfiniteKeyLoader } from 'swr/infinite'

const keyLoader: SWRInfiniteKeyLoader = (index, previousPageData) => { /* ... */ };
const useList = () => {
  const { data } = useSWRInfinite(keyLoader, fetcher);
  // ...
};
```

### Generics

> Speicifying the type of `data` is easy.

Reviewer 2: ```suggestion
Speicifying the type of `data` is easy. By default, it will use the return type of `fetcher` (with `undefined` for the non-ready state) as the `data` type, but you can also pass it as a parameter:
```

```typescript

> const { data } = useSWR<string[]>('/api/data', fetch, config)

Reviewer 2: ```suggestion
// 🔹 A. Use a typed fetcher:
// `getUser` is `(endpoint: string) => User`.
const { data } = useSWR('/api/user', getUser)

// 🔹 B. Specify the data type:
// `fetcher` is generally returning `any`.
const { data } = useSWR<User>('/api/user', fetcher)
```

In both cases, `data` will be typed as `User | undefined` automatically.

```

> If you wanna type for all the arguments of useSWR you can also import them from `swr`:

Reviewer 2: ```suggestion
If you want to add types for other options of SWR, you can also import those types directly:
```

> ```js

Reviewer: ```suggetion
typescript
```

import { useSWR } from 'swr'
import type { SWRConfiguration } from 'swr'

const config: SWRConfiguration = {
  fallbackData: "fallback",
  revalidateOnMount: false
  // ...
}

const { data } = useSWR<string[]>('/api/data', fetcher, config)
```

### Middleware Types

> 

Reviewer 2: ```suggestion

There're some extra type definitions you can import to help adding types to your custom middleware.

```

# TypeScript

SWR is friendly for apps written in TypeScript, with type safety out of the box.

### Basic Usage

By default, SWR will also infer the argument types of `fetcher` from `key`, so you can have the preferred types automatically.

#### useSWR

```typescript
// `key` is inferred to be `string`
useSWR('/api/user', key => {})
useSWR(() => '/api/user', key => {})

> // `key` will be inferred as { a: string; b: { c: string; d: number } }

Reviewer 2: ```suggestion

// `key` will be inferred as { a: string; b: { c: string; d: number } }
```

So it won't be confused with previous lines.

useSWR({ a: '1', b: { c: '3', d: 2 } }, key => {})
useSWR(() => { a: '1', b: { c: '3', d: 2 } }, key => {})

> // `arg0` will be inferred as string.  `arg1` will be inferred as number

Reviewer 2: ```suggestion

// `arg0` will be inferred as string.  `arg1` will be inferred as number
```
